Add Kafka raw encoder

charlesjmorgan opened the pull request:
Add `RawRowEncoder` and `RawRowEncoderFactory`
Add test case in `io.prestosql.plugin.kafka.TestKafkaIntegrationSmokeTest#testRoundTripAllFormats`

losipiuk commented:
Make commit message just `Add Kafka raw encoder`. The fact that tests come together with implementation is natural.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,282 @@
+        if (!mappingMatcher.matches()) {
+            throw new IllegalArgumentException(format("Invalid mapping format '%s' for column '%s'", mapping, columnHandle.getName()));
+        }
+        int start = parseInt(mappingMatcher.group(1));
```
nit: possibly catch `NumberFormatException` and wrap in exception with clear message. It can throw if value provided does not fit in int.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,282 @@
+        }
+    }
+
+    private int parseMapping(EncoderColumnHandle columnHandle, FieldType fieldType)
```
can we make it `@VisibleForTesting` and add unit test for it?

charlesjmorgan replied:
I could test it without `@VisibleForTesting` annotation just by calling encoder factory create method on columns with different mappings

charlesjmorgan replied:
but that wouldn't actually test the method

losipiuk replied:
Given logic complexity I would probably opt toward directly unit testing method. But if you can make it look nice and readable testing at wider scope. And still be able to excercise all corner cases, it should be fine too.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,282 @@
+
+        int length = end.isPresent() ? end.getAsInt() - start : fieldType.getSize();
+
+        if (!isVarcharType(columnHandle.getType())) {
```
it does not seem to handle case where we are dealing with varchar and `end` is not explicitly specified.
Would lenght be `1` then?

charlesjmorgan replied:
The way I had the encoders structured before didn't let me 'see' the value of the mapping for the next column so `varchar` types would need to specify the start and the end every time. Now I think I might be able to dynamically allocate space for varchar types by 'peaking' into the mapping for the next column and using the `start index - 1` for the next column as the end index for the varchar column. I could also use `VarcharType#getLength` if there isn't a next column.

charlesjmorgan replied:
that's how the decoder works rn afaik

aalbu replied:
So that means that a missing end value is legal only when the field is in the last position.

charlesjmorgan replied:
Yes, I will have to figure out a way to allocate more space in the `ByteBuffer` if `end` is not specified for `VARCHAR`.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,282 @@
+    @Override
+    protected void appendString(String value)
+    {
+        buffer.put(value.getBytes(StandardCharsets.UTF_8), currentBufferPosition, valueLengths[currentColumnIndex]);
```
this can throw `value.getBytes()` may return shorter array than `valueLengths[currentColumnIndex]`

charlesjmorgan replied:
the best solution I can think of is create an intermediate `ByteBuffer` of the correct length for that column and then put the array from that into the actual `ByteBuffer` like this
```java
protected void appendString(String value)
{
    ByteBuffer tempBuf = ByteBuffer.allocate(valueLengths[currentColumnIndex]);
    tempBuf.put(value.getBytes(StandardCharsets.UTF_8), 0, valueLengths[currentColumnIndex]);
    buffer.put(tempBuf.array(), currentBufferPosition, valueLengths[currentColumnIndex]);
    currentBufferPosition += valueLengths[currentColumnIndex];
}
```

charlesjmorgan replied:
Yes for now I think that would be best. In the future it would be a good idea to rethink how the raw encoder maps values and better define that in the documentation.

losipiuk replied:
Actually I meant that the code will throw if `String value` is shorter than `valueLengths[currentColumnIndex]`. It is still the case with updated code.
`buffer.put(srcArray, offset, length)` requires that there is at least `length` bytes in `srcArray` starting from `offset`.

Also I have just noticed that you pass `currentBufferPosition` source offset. This does feels wrong as the offset should be in source string space. And `currentBufferPosition` represents position in target buffer. You should pass 0 as offset i think.

charlesjmorgan replied:
> Actually I meant that the code will throw if `String value` is shorter than `valueLengths[currentColumnIndex]`
In this case I think that it is best if we throw an exception for now, otherwise the decoder will just read garbage.
> Also I have just noticed that you pass `currentBufferPosition` source offset. This does feels wrong as the offset should be in source string space.
I noticed this as well while working on this issue last week. I also realized that `currentBufferPosition` is a redundant variable because I can just use the built in `position` feature of the `ByteBuffer` to track the position and then reset it in `toByteArray`.

losipiuk replied:
>> Actually I meant that the code will throw if String value is shorter than valueLengths[currentColumnIndex]

> In this case I think that it is best if we throw an exception for now, otherwise the decoder will just read garbage.

Yeah. I think I agree. Unless we want to pad with some whitespace. But still it should be a descriptive exception. Not IndexOutOfBounds... :)

charlesjmorgan replied:
what if it's `IndexOutOfBoundsException` with a descriptive message?

losipiuk replied:
That should be fine :)

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,282 @@
+
+        checkArgument(start >= 0, "Start offset %s for column '%s' must be greater or equal 0", start, columnHandle.getName());
+        end.ifPresent(endValue -> {
+            checkArgument(endValue >= 0, "End offset %s for column '%s' must be greater or equal 0", endValue, columnHandle.getName());
```
Right, with the changes I just made I think that it is now necessary

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,334 @@
+        }
+    }
+
+    private int parseMapping(int index, FieldType fieldType)
```
I'll try this out, already sounds much easier than what I was doing.

charlesjmorgan replied:
This still leaves the question of what to do with `VARCHAR` types that don't have an end specified. The way I see it we could,

1. Set the length equal to 1 (byte)
2. Guess the length based on the start mapping of the next column
3. Guess the length based on the length of the `VARCHAR` column type instance

Otherwise I don't know how many bytes to allocate to the `ByteBuffer`

losipiuk replied:
I think that if INSERT is to be supported for table we should require `end` to be specified for VARCHAR columns. Possibly allowing only exception where VARCHAR column maps to tail of message byte buffer; I am not sure if that is necessary though.
I think this is fine that INSERT not necessarily works for all table definitions for which SELECT works (overlapping fields is a good example where we cannot make INSERT work). 
Let's add constraints and focus on on basic scenarios first (with well defined semantics). Also let's make sure that for those, INSERT is compatible with SELECT (we can do roundtrip). If we see that it would be useful, we can relax contraints later.

aalbu replied:
Based on reading the docs and skimming through the code, it seems to me that a `VARCHAR` mapping without an `end` specified is valid **only** for the last column in the mapping.  For any other position, I think we need the `end` index and we need to ensure that the length of the value we are inserting matches the size of the mapping.

I agree about the importance of round-trip tests, but we also need to make sure we conform to the spec we created originally, as up to this point, data has been created by non-Presto clients.

charlesjmorgan commented:
I think I simplified/separated the mapping logic, let me know if what I did works better/worse.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,375 @@
+
+        public int getLength()
+        {
+            if (end.isPresent()) {
```
I do not believe you are calling this method ever with end non-set. Add precondition that `end.isPresent()` and drop `if`.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,375 @@
+                        }
+                    }
+                    else {
+                        this.end = OptionalInt.empty();
```
I would initialize this.end with `OptionalInt.empty()` and drop `else` sth like:
```java
                    this.start = parseInt(mappingMatcher.group(1));
                    this.end = OptionalInt.empty();
                    if (mappingMatcher.group(2) != null) {
...
```

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,375 @@
+                    throw new IllegalArgumentException(format("Invalid mapping for column '%s'", this.name));
+                }
+                else {
+                    this.start = parseInt(mappingMatcher.group(1));
```
wrap with `try`

Maybe extract method `parseOffset()` which does exception handling and call it like:
```
parseOffset(mappingMatcher.group(1), "start", this.name)
parseOffset(mappingMatcher.group(2), "end", this.name)
```

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,375 @@
+                valueBytes.length,
+                columnHandles.get(currentColumnIndex),
+                columnMappings.get(currentColumnIndex).getLength()));
+        for (int i = 0; i < columnMappings.get(currentColumnIndex).getLength(); i++) {
+            buffer.put(valueBytes[i]);
+        }
```
```suggestion
        buffer.put(valueBytes, 0, valueBytes.length)
```

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,375 @@
+    protected void appendString(String value)
+    {
+        byte[] valueBytes = value.getBytes(StandardCharsets.UTF_8);
+        checkArgument(valueBytes.length == columnMappings.get(currentColumnIndex).getLength(), format(
```
This is a strong constraint. Maybe allowing for shorter strings with some user-defined padding character would give better ux?
@findepi @aalbu WDYT? (can be relaxed as a followup)

charlesjmorgan replied:
I didn't want to change the specifications for the raw format in this pr. I agree with you that padding shorter strings would be a good idea. The specifications for the raw format need to be rethought and clearly defined in the docs, figured it would be best to do that in a follow-up pr.

aalbu replied:
I agree with @charlesjmorgan.  Without being explicit in the specification, it's probably best to require the users to pad in the query.

losipiuk commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -149,6 +149,10 @@ public void testRoundTripAllFormats(RoundTripTestCase testCase)
                         "all_datatypes_csv",
                         ImmutableList.of("f_bigint", "f_int", "f_smallint", "f_tinyint", "f_double", "f_boolean", "f_varchar"),
                         ImmutableList.of(100000, 1000, 100, 10, 1000.001, true, "'test'")))
+                .add(new RoundTripTestCase(
```
I know that internally `RoundTripTestCase` now sends and receives two rows. But IMO it would be nicer if each row would be different, not same one repeated twice. Can we improve on that?

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,375 @@
+                    columnMapping.setEnd();
+                }
+                else {
+                    if (this.columnMappings.indexOf(columnMapping) == this.columnMappings.size() - 1) {
```
Yeah that looks better to me

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,375 @@
+                if (!isVarcharType(columnMapping.getType())) {
+                    columnMapping.setEnd();
+                }
+                else {
```
so the logic for this for loop should look like this?
```java
if (mapping.getEnd().isPresent()) {
    continue;
}
else if (!isVarcharType(mapping.getType())) {
    mapping.setEnd(mapping.getStart() + mapping.getFieldType().getSize());
}
else if (mapping.getName().equals(getLast(columnHandles).getName())) {
    mapping.setEnd(mapping.getStart() + mapping.getFieldType().getSize());
}
else {
    throw new IndexOutOfBoundsException(format(
            "No end mapping defined for column '%s' of type '%s'",
            mapping.getName(),
            mapping.getType().getDisplayName()));
}
```

aalbu replied:
You don't need the `else` after the `continue`.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,345 @@
+        for (EncoderColumnHandle handle : this.columnHandles) {
+            columnMappingsBuilder.add(new ColumnMapping(handle));
+        }
+        this.columnMappings = columnMappingsBuilder.build();
```
A functional approach would make this shorter and more readable.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,345 @@
+                if (!mappingMatcher.matches()) {
+                    throw new IllegalArgumentException(format("Invalid mapping for column '%s'", this.name));
+                }
+                else if (mappingMatcher.group(2) != null) {
```
`else` is unnecessary.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,345 @@
+                    try {
+                        this.end = parseOffset(mappingMatcher.group(2), "end", this.name);
+                    }
+                    catch (NumberFormatException e) {
```
It doesn't look like the method throws `NumberFormatException`.  I also don't get the exception message.  But it doesn't look like you need to catch here.

charlesjmorgan replied:
oh right, I moved the error handling to the method and forgot to remove it here

charlesjmorgan commented:
- Added a separate commit to add different rows to the round trip tests. I moved some helper methods into `ColumnMapping`.
- Tried using a `ColumnMappingBuilder` class to create `ColumnMapping's` but realized that it would be easier to just change the type of `end` from `OptionalInt` to `int`. This way end always has a value and it's up to the `RawRowEncoder` constructor to make sure that the mappings for the columns are valid.
- Made `ColumnMapping` static and parameterized the constructor with `EncoderColumnHandle`.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,345 @@
+    protected void appendNullValue()
+    {
+        for (int i = 0; i < columnMappings.get(currentColumnIndex).getLength(); i++) {
+            buffer.put((byte) 0);
```
This is another missing part in the spec, right?  It doesn't say how nulls should be encoded.

charlesjmorgan replied:
Correct. I could be interpreting this wrong but based on the code in the `RawColumnDecoder` it looks like the only case in which a column can be null is if `start` and `end` values for the mapping are equal. Guess I should remove `appendNullValue` for now and we can figure that out in a future commit.

charlesjmorgan replied:
Ugh the append null method is used when no key value is specified

losipiuk replied:
> when no key value is specified

What do you mean by that. INSERT when only subset of columns are specified?

Can we just disallow writing NULLs into table with Raw encoders? Which usecases does it break?
It seems that with mapping we now support for write we would not be able to write a value which would be interpreted as NULL by decoder anyway.

charlesjmorgan replied:
It wouldn't necessarily break anything, but you would have to insert values for every column when inserting. This includes the key column(s). It would definitely be best to disable writing null values until we can figure out a way to properly read/write them.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,357 @@
+        checkArgument(currentColumnIndex == columnHandles.size(), format("Missing %d columns", columnHandles.size() - currentColumnIndex + 1));
+
+        resetColumnIndex(); // reset currentColumnIndex to prepare for next row
+        buffer.clear(); // reset buffer position to prepare for next row
```
first `clear()` and then `array()`. I am supprised it works. I think we should just mark intention to `clear` here and do actual clearing on next append.

charlesjmorgan replied:
`clear()` doesn't actually get rid of any of the data in the backing array, it just resets the position and limit

losipiuk replied:
Yeah. You are right. I was expecting that this is just imlementation detail but actually it is documented behaviour. Can you make the comment by `buffer.clear()` more explicit so it is obvious that we know what we are doing, and avoid confustion of next generation of people reading this code :)

charlesjmorgan replied:
yeah for sure

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,344 @@
+    // parse column mappings from column handles
+    private static List<ColumnMapping> buildMappings(List<EncoderColumnHandle> columnHandles)
+    {
+        return columnHandles.stream().map(ColumnMapping::new).collect(toImmutableList());
```
You can inline this.

charlesjmorgan replied:
think I misunderstood what you meant by a functional approach originally hahah

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,344 @@
+                mapping.getStart(),
+                mapping.getName(),
+                expectedPosition));
+        checkArgument(mapping.getEnd() > mapping.getStart(), format(
```
`checkMappingOffsets()` is validating (in part) that the encoder itself is constructed properly.  The constructor of `ColumnMapping` is responsible for making sure that those objects are properly constructed.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java:
```diff
@@ -0,0 +1,344 @@
+            catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException(format("Invalid dataFormat '%s' for column '%s'", dataFormat, columnName));
+            }
+            return FieldType.BYTE;
```
Can you move this statement right after the `if`?

losipiuk commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -144,62 +144,64 @@ public void testRoundTripAllFormats(RoundTripTestCase testCase)
-            return fieldValues.get(index);
+            String[] rows = new String[numRows];
+            for (int i = 0; i < numRows; i++) {
+                rows[i] = "(" + rowValues.get(i).stream().map(Object::toString).collect(Collectors.joining(", ")) + ")";
```
```suggestion
                rows[i] = rowValues.get(i).stream().map(Object::toString).collect(Collectors.joining(", ", "(", ")"));
```
